### ISSUE TITLE
Creating an OSARA build for Linux?

### ISSUE THREAD

**Issue by ethindp** [jcsteh/osara#176]
Hello guys,
I'm curious if it would be possible to begin a project for OSARA to be used on Linux. As you guys no doubt know, Reaper has released experimental builds for Linux. The problem is that you don't get (so far as I know) OSARA and SWS&SM on there, so for the visually impaired, using Reaper in general would be a challenge. I'd be happy to contribute and get the ball rolling, as it were, but I haven't done very much Linux programming, so don't know the equivalent API calls. If someone could either get the project going or give me resources and help me out with the project, that would be great. While we won't have SWS&SM on Linux if this works, we'll at least have OSARA, which would be a huge step towards music production and audio editing with Reaper on Linux.

**Comment by jcsteh** [comment]
I personally don't have time to even look into this, so unless someone else wants to do the research and implementation, this is unfortunately a won't fix.

REAPER (and OSARA) uses a Cockos library called SWELL to emulate Win32 GUI stuff on other platforms. I don't know what widgets that uses on Linux. If it uses widgets that already have accessibility (e.g. GTK 2), that hopefully means most REAPER and OSARA menus, dialogs, etc. will be accessible. I'd start by checking whether REAPER's menus and dialogs are accessible already. If they aren't, there's no point in even starting to port OSARA. You'd need to work with Cockos to get the GUI issues fixed first.

REAPER also uses some custom widgets (e.g. for faders), and WDL wingui (another Cockos library) provides specific accessibility support on both Windows and Mac. I'm not sure whether it provides accessibility support on Linux. If it doesn't, again, porting OSARA won't fix that; you'd need to work with Cockos to get that fixed.

You'll also need to write a Linux implementation of the OSARA code for reporting messages. I guess you might be able to manage this by implementing a live region using your own ATK accessible object or something.

I'm closing this as won't fix because I can't work on this and I very much doubt anyone else will kick it off either. If someone comes along and does decide to start real work on this, we can always reopen.

**Comment by leonardder** [comment]
I put some research into this as it sounds promising that you can run OSARA on a raspberry PI. It should also be possible to build a version of Swell based on GTK2.

**Comment by leonardder** [comment]
@jcsteh Would you be open to reopening this to investigate further? I"m currently hooking up a vm to test how it behaves by default, and will look into GTK2 if the default GTK3 setup fails.

**Comment by leonardder** [comment]
Ah never mind, I scrambled GTK with GDK. Furthermore, I tried both the default swell as well as a custom compiled GDK2 variant, without success. OCR shows me that the Reaper window is shown correctly, but Orca doesn't give a shout at all.

**Comment by leonardder** [comment]
While this currently can't be done because REAPER is completely inaccessible on Linux, I'm reopening this and giving this an appropriate label to point out that this requires work by Cockos.

**Comment by jcsteh** [comment]
@leonardder, is this something you are planning to pursue with Cockos? If not, I think we should keep this closed as won't fix because I don't see anyone else coming along to work on it and I don't want to leave anyone with false hope that this will ever get any attention. We can always reopen if someone wants to pursue it.

**Comment by leonardder** [comment]
I'm really interested and motivated getting this to work myself.

If @ScottChesworth  could use  his contacts with Cockos to point out general Linux inaccessibility, that would be really helpful. In the meanwhile, there are still some experiments I'd like to take myself, such as trying to find out how accessibility is implemented in Swell for Mac, and how we could make a Swell based app accessible on Linux. If that can be done without too much efford, we can even recompile swell and plug that into a Cockos REAPER build, as they even suggest themselves.

**Comment by jcsteh** [comment]
> I'm really interested and motivated getting this to work myself.

Fair enough.

> If @ScottChesworth could use his contacts with Cockos to point out general Linux inaccessibility, that would be really helpful.

IMO, for a request this big, it's going to take a lot of technical discussion to get it resolved and it doesn't make sense for someone to be an intermediary. Given your technical background and competence, I would just contact Cockos yourself via email (support@cockos.com). That's how I established contact with them and probably how Scott did also.

> In the meanwhile, there are still some experiments I'd like to take myself, such as trying to find out how accessibility is implemented in Swell for Mac,

As I understand it, Swell emulates Win32 calls and translates them to native widgets on Mac. REAPER also uses custom virtwnd controls (see WDL/wingui), for which there are both Windows IAccessible and Mac a11y implementations.

**Comment by ScottChesworth** [comment]
> IMO, for a request this big, it's going to take a lot of technical discussion to get it resolved and it doesn't make sense for someone to be an intermediary.
> 
Agreed. I don't know enough about the topic to lead that conversation.
> I would just contact Cockos yourself via email
> 
I'm happy to be CC'd in on that, can keep an eye on the conversation and nudge periodically to keep things moving along if that would be helpful.
BTW @leonardder, when you make contact, mention "OSARA" in the subject line so they know who you roll with.

**Comment by jcsteh** [comment]
I'm happy to be on CC as well if that is helpful. I can't promise participation, but I do know a little about the Linux a11y API at least from my work on Firefox, so I can potentially chip in if there's a question I can help answer.

**Comment by leonardder** [comment]
I contacted Cockos with some help from @ScottChesworth , and I"m happy to say that they're really open to add Linux accessibility to swell, though it's not yet explicitly on their roadmap and they don't really know where to start.
As far as I've read, adding accessibility to something GTK4 based should be much easier than adding it to the older GTK versions, so I asked them whether there's anything like supporting GDK4/GTK+4 on their map yet.

**Comment by jcsteh** [comment]
I was involved in a [Twitter discussion](https://twitter.com/jcsteh/status/1422479201865637897) with Justin from Cockos about this. Long story short: for various reasons, REAPER draws its own widgets on Linux rather than using mostly native widgets like it does on Windows and Mac. That means they'd have to implement accessibility APIs from scratch for the GUI. I did some digging and pointed them at some starting material for that, though.

**Comment by ethindp** [comment]
So, I know this issue is old, but the landscape has, I think, radically changed.

First, I have been, for the last few months, developing [Prism](https://github.com/ethindp/prism) which is a universal TTS/screen reader abstraction layer across all the platforms I am aware of. Although it uses C++23, this is only for the (internal) library itself, and C23 is not required to interact with it, since the API is clean C. The library is still early in development and changes I'm sure will definitely need to be made, but from what I know it's quite stable now and the problems that are encountered are usually just issues with back-ends, which only require backend-local changes, limiting the blast radius significantly.

The second is that AccessKit is now a thing and is very heavily used in things like Godot. As such, I imagine it would dramatically simplify Cocko's attempts to make SWELL accessible, and the process is just adding it to all the widgets and then ironing out problems to make the UI feel as "native" as possible (and this is the part I sadly have little experience with). In theory, you could just use AccessKit for everything, but Prism and AccessKit make quite the potent combination (although I'm quite biased, so...).

I wonder if both of the aforementioned might make all of this dramatically easier?

**Comment by RDMurray** [comment]
I started vibe coding accesskit integration for swell last night. there's a very incomplete proof of concept at https://github.com/RDMurray/WDL/tree/accesskit

It does make various controls like buttons and edit controls accessible and works in Reaper by replacing libSwell.so.

So yes, it's definitely possible. I wonder how many people are interested in Reaper + OSARA on Linux?

**Comment by ethindp** [comment]
@RDMurray I am definitely interested and I know of several others that are as well. Reaper, fully accessible, on Linux, would be really neat.

**Comment by ScottChesworth** [comment]
There was a discussion about it on RWP a couple of months back. To summarize, a handful of people seemed keen as users, nobody with technical chops popped up expressing eagerness to do the gruntwork involved. If you wanna do that and you can be present on the groups to answer Linuxy things, knock yourself out, I'm sure some folks would use it. Can't really say I'd be one of them right now, not enough plug-in support over there yet for me.

**Comment by RDMurray** [comment]
Okay I'll keep going with the swell accessibility on Linux and see where it gets to.  I think we'll have to make it clear that it isn't supported in the usual groups, at least for a while. I honestly haven't tried anything on desktop Linux for a long time, but anything to do with audio used to be a complete nightmare.

Is the situation improved when using Jack for audio apps along with a screen reader? 

I'm going to have to install Linux on an old laptop for testing this. I'm developing in WSL for now.

For all I know the code could be a complete mess, it's 100% done in Codex. currently over 4000 lines - there's no way I'm ever manually reviewing that, so it's unlikely to be merged by Cockos.

**Comment by ethindp** [comment]
@RDMurray Pipewire is the primary audio system for Linux nowadays, and it's much better. It emulates Jack, Pulse, all the big audio back-ends that things used to use, and is much better, I think anyway, than the originals.

As for it getting merged? Yeah, I don't think Cockos would want AI-generated code in that particular repo. Or maybe they would, would be worth asking.

**Comment by RDMurray** [comment]
That's great to hear. MS have somehow managed to make audio almost completely unusable in WSL. 

It might not matter, as long as swell is dynamically linked we can maintain a fork. I'll let you know when there's a build for testing. It's still very buggy and incomplete at the moment.

**Comment by RDMurray** [comment]
see #1398  for Linux support. please test and report any problems. I am assuming Linux knowledge for now. There is no installer. you need to replace `libSwell.dll` included with reaaper with the accessible version at https://github.com/RDMurray/WDL

extract the osara tarball in the reaper resource path, default is `~/.config/REAPER`. the keymap is in `KeyMaps/OSARA.ReaperKeyMap`, manually copy it to `reaper-kb.ini`

The keymap is just a copy of the windows keymap for now, there may be conflicts.

Feedback appreciated.

**Comment by m1maker** [comment]
@RDMurray

Hello

Since Linux is my main OS, I'm very interested in an accessible DAW. This is very good news, because I can return to creativity. Thank you!

I've tested both SWELL and OSARA, and I can tell you that it really works!
There are some issues that I would write about in GitHub on your WDL fork if they were allowed.

I will definitely take the time to read it's code. Maybe I can suggest/contribute something.
But in general, the screen reader is announcing most of GUI, which is good.

**Comment by m1maker** [comment]
@RDMurray 

So. I used REAPER, read the code, and I can say the following:
Firstly, the slider (aka `REAPERhfader`) values are not readable.
The fact is that this is a custom control, and in order to get its values, we need a different approach.
```C++
static void swell_accesskit_fill_numeric_state(HWND hwnd, SWELL_AccessKitOwnedNode *node)
{
  if (!hwnd || !node || !hwnd->m_private_data || !hwnd->m_classname) return;
```

Stop here, because the third condition will never be true on this control because `m_private_data` is always 0.
To correctly implement accessibility for at least this object, look:
`WDL/wingui/virtwnd-iaccessible.cpp`
&&
`WDL/wingui/virtwnd-nsaccessibility.mm`

Cockos has a whole framework for their own widgets, for which they have implemented accessibility on Windows and macOS.
Of course, I understand that the AccessKit model is different, but it can all be rewritten, that's not the main thing. And the main thing is that here we have access to these internal structs/method declarations of these virtual windows, and therefore we can correctly get all the values.
Maybe this can be done in `WDL/swell/swell-accesskit-generic.cpp` but I strongly doubt it, because it is unlikely that it will be possible to cast `HWND` to this internal virtual window.
I tried to get slider values via `SendMessage` with type `MB_GETPOS`, but it was giving out some too high values.
I'll come back to this issue later, maybe I can help with the implementation. I also need some time to understand all this architecture in more detail.

**Comment by RDMurray** [comment]
I gave up on trying to make custom controls in wingui accessible for now. Wingui is statically linked into Reaper, and symbols are stripped so I can't see any way to introspect those controls. Maybe we can ask them to dynamically link wingui as well.

Issues and pull requests are enabled on RDMurray/wdl now.

**Comment by LeonarddeR** [comment]
Now this is fun. I played an evening with exactly the same idea and prototyped a Swell change with fable at [my own fork](https://github.com/LeonarddeR/WDL/tree/linuxA11y), only to discover @RDMurray beat me to it. Congratulations on this work, I"m really motivated to assist where possible.

My approach is fundamentally different because I decided not to go the accesskit route but use atspi directly, because figured it would be more portable and therefore easier to upstream for Cockos.

Also note that it is not trivial, but certainly possible to test this on WSL.

**Comment by ethindp** [comment]
I feel like accesskit is probably the better route (since we can then improve accesskit as this effort progresses). But maybe atspi is easier to integrate? Not sure.

**Comment by RDMurray** [comment]
@LeonarddeR did you manage to get reasonable performance using atspi directly? I started a branch for that as well but it was so slow. Of course that's because I didn't implement whatever optimisations that accesskit already does and I let the AI loose without learning much about how atspi is supposed to be implemented.

I started on wsl, but switched to Fedora because I couldn't get  enough of the Orca and Reaper keyboard shortcuts to work.

**Comment by LeonarddeR** [comment]
> [@LeonarddeR](https://github.com/LeonarddeR) did you manage to get reasonable performance using atspi directly? I started a branch for that as well but it was so slow. 

Yes it was very slow in first instance at my end too, but after an hour of Fable debugging it was much faster. The improvements related to that are in the more recent commits on my branch.

@ethindp I agree that for the open source blind community, accesskit is the way to go, but then we can most likely forget COCockos will ever consider upstreaming it.

**Comment by RDMurray** [comment]
@LeonarddeR it's probably worth keeping going with your fork anyway - we can compare results and it would be really good if Cockos could upstream it because that's the only way we could get a11y for their wingui widgets unless they change to dynamically linking wingui.

**Comment by m1maker** [comment]
Yes, if they linked wingui dynamically, it would help us a lot. Because I fixed slider announcements but it's a very dirty hack that Cockos will never merge. By the way, they don't have some of the window messages that Microsoft has (more details [here](https://github.com/rDMurray/wDL/pull/1)).

I'm already questioning whether this pool request should be merged into your fork, but I'm still using it. The only thing is that the values are sometimes strange, for example, VolumeDB is equal to 700, then 716 when incremented, etc.

By the way, there's an issue with the plugin tree, but I haven't looked into why it only announced once when focusing.

 Additionally, has anyone contacted Cockos regarding linking?

@LeonarddeR
I will also test your fork.

**Comment by pvagner** [comment]
I am interested in this as well.

This morning I have found yet another attempt at wibecoding this. it also uses at-spi-directly but it does everything at osara side of things without forking WDL. It's needed for building osara though. It's at https://github.com/churst90/osara-linux

I am using evaluation version of reaper 7.79 on arch linux.
In order to make this version of reaper actually load forked WDL I had to rebase both @LeonarddeR's and @RDMurray's fork against the upstream.

In terms of completeness the accesskit based one by @RDMurray seems to be most advanced. It reads menubar including closed menus, roles, states, mnemonics and similar. However whenever focus changes orca is reading the title of the active window. And it does not read focus changes inside dialogs such as when using tab or shift+tab to navigate in the preferences window. Menus, treeview in settings, osara prompts such as when navigating in the list of tracks or using various toggles read fine though.

Then the at-spi based WDL fork by @LeonarddeR is working great inside the dialogs but it does not report parrent menubar menus. When navigating in the menus it causes orca to report role with each focus change. I don't know how to build osara so it's working alongside @LeonarddeR's WDL fork yet though.

For me as a novice to reaper building the one by @churst90 was most easiest. It has same issue with menus like the WDL fork by @LeonarddeR however I can get the matching osara build working easily.
